building actonc fails after source code changes

### Acton Version

0.11.7.20221014.9.24.53

### Steps to Reproduce

This behaviour is new for me since I got a MacBook Air M1 running MacOS 12.5.1

Change some Haskell source code in the compiler and do make


### Expected Behavior

rebuilding of the system

### Actual Behavior

Building executable 'actonc' for acton-0.11.7.20221014.9.24.53..
[29 of 30] Compiling Paths_acton
[30 of 30] Compiling Main
Linking .stack-work/dist/aarch64-osx/Cabal-3.2.1.0/build/actonc/actonc ...
acton> copy/register
Installing executable actonc in /Users/sydow/acton/compiler/.stack-work/install/aarch64-osx/c0a61ca3b9d848a424af2019a052e93294f3ff541273ae429a5b30cf4ff52996/8.10.7/bin
cp compiler/actonc dist/bin/actonc
dist/bin/actonc builtin/ty/src/__builtin__.act
make: *** [builtin/ty/out/types/__builtin__.ty] Killed: 9


The problem is probably in 
cp compiler/actonc dist/bin/actonc

After the above failure, I can run compiler/actonc, but dist/bin/actonc crashes, in spite of the files being identical.

If I now *move* compiler/actonc to dist/bin, everything works. I don't understand this, but googling gives several hits with similar behaviour in MacOS. 

Suggested solution:
Change cp to mv in line 424 of acton/Makefile (or, if one for some reason I don't understand wants to keep compiler/makefile, use the same hack as for actondb a few lines further down).

So I picked up my M2 and reproduced the problem. It doesn't just fail temporarily, I cannot run this actonc afterwards, as a permanent error. How is this possible? What is it doing?

```
➜  acton-m2 git:(main) ✗ make -j && dist/bin/actonc examples/helloworld.act&& examples/helloworld
cd compiler && stack build --dry-run 2>&1 | grep "Nothing to build" || \
		(sed 's,^version:.*,version:      "0.11.7.20221016.1.8.17",' < package.yaml.in > package.yaml \
		&& stack build --ghc-options -j4 \
		&& stack --local-bin-path=. install 2>/dev/null)
acton> configure (exe)
Configuring acton-0.11.7.20221016.1.8.17...
acton> build (exe)
Preprocessing executable 'actonc' for acton-0.11.7.20221016.1.8.17..
Building executable 'actonc' for acton-0.11.7.20221016.1.8.17..
[29 of 30] Compiling Paths_acton
[30 of 30] Compiling Main
Linking .stack-work/dist/aarch64-osx/Cabal-3.2.1.0/build/actonc/actonc ...
acton> copy/register
Installing executable actonc in /Users/kll/terastream/acton-m2/compiler/.stack-work/install/aarch64-osx/892a08223c78b0d033cd52a84638c717ed71d132975dfdc30a8c5c41041aa667/8.10.7/bin
cp compiler/actonc dist/bin/actonc
dist/bin/actonc builtin/ty/src/__builtin__.act
make: *** [builtin/ty/out/types/__builtin__.ty] Killed: 9
➜  acton-m2 git:(main) ✗ make -j && dist/bin/actonc examples/helloworld.act&& examples/helloworld
/bin/sh: line 1: 24625 Killed: 9               dist/bin/actonc --numeric-version 2> /dev/null
/bin/sh: line 1: 24631 Killed: 9               dist/bin/actonc --numeric-version 2> /dev/null
dist/bin/actonc builtin/ty/src/__builtin__.act
make: *** [builtin/ty/out/types/__builtin__.ty] Killed: 9
➜  acton-m2 git:(main) ✗ md5sum dist/bin/actonc compiler/actonc
6ecbc2066fef459040a42b67abe1b2ca  dist/bin/actonc
6ecbc2066fef459040a42b67abe1b2ca  compiler/actonc
➜  acton-m2 git:(main) ✗ compiler/actonc
Missing: CMD

Usage: actonc [--numeric-version] [--version] [--parse] [--kinds] [--types]
              [--sigs] [--norm] [--deact] [--cps] [--llift] [--hgen] [--cgen]
              [--ccmd] [--quiet] [--verbose] [--debug] [--timing] [--stub]
              [--dev] [--cpedantic] [--tempdir TEMPDIR] [--syspath TARGETDIR]
              [--root ARG] CMD [ARG]
  Compile an Acton source file with recompilation of imported modules as needed
➜  acton-m2 git:(main) ✗ dist/bin/actonc
[1]    24690 killed     dist/bin/actonc
➜  acton-m2 git:(main) ✗ cp compiler/actonc dist/bin/actonc
➜  acton-m2 git:(main) ✗ dist/bin/actonc
[1]    25047 killed     dist/bin/actonc
➜  acton-m2 git:(main) ✗ rm dist/bin/actonc
➜  acton-m2 git:(main) ✗ cp compiler/actonc dist/bin/actonc
➜  acton-m2 git:(main) ✗ compiler/actonc
Missing: CMD

Usage: actonc [--numeric-version] [--version] [--parse] [--kinds] [--types]
              [--sigs] [--norm] [--deact] [--cps] [--llift] [--hgen] [--cgen]
              [--ccmd] [--quiet] [--verbose] [--debug] [--timing] [--stub]
              [--dev] [--cpedantic] [--tempdir TEMPDIR] [--syspath TARGETDIR]
              [--root ARG] CMD [ARG]
  Compile an Acton source file with recompilation of imported modules as needed
```

What is this sorcery?

--- a/Makefile
+++ b/Makefile
@@ -419,15 +419,16 @@ clean-rts:
 # == DIST ==
 #
 
-$(ACTONC): compiler/actonc
-	@mkdir -p $(dir $@)
-	cp $< $@
-
 # This does a little hack, first copying and then moving the file in place. This
-# is to avoid an error if actondb is currently running. cp tries to open the
-# file and modify it, which the Linux kernel (and perhaps others?) will prevent
-# if the file to be modified is an executable program that is currently running.
-# We work around it by moving / renaming the file in place instead!
+# is to avoid an error if the executable is currently running. cp tries to open
+# the file and modify it, which the Linux kernel (and perhaps others?) will
+# prevent if the file to be modified is an executable program that is currently
+# running.  We work around it by moving / renaming the file in place instead!
+dist/bin/actonc: compiler/actonc
+	@mkdir -p $(dir $@)
+	cp $< $@.tmp
+	mv $@.tmp $@
+
 dist/bin/actondb: backend/actondb
 	@mkdir -p $(dir $@)
 	cp $< $@.tmp

